rfcs/20180817-variables-20.md

```diff
@@ -0,0 +1,140 @@
+    *   users will be able to implement their own variable-like objects by subclassing tf.Variable and adding a scoped factory function to use those variables
+*   variable_scope and get_variable will be removed
+    *   the tf 1.0 version of variable_scope and get_variable will be left in tf.compat.v1
+    *   to control variable naming users can use tf.name_scope + tf.Variable
+    *   whether a variable is shared across sessions / processes will be controlled by a constructor argument to tf.Variable; no other type of scope reuse will be done in the framework
+    *   scoped partitioning will be implemented as a factory function at first
+    *   libraries and users are encouraged to reuse variables by reusing their objects, like Keras layers do
+    *   custom_getters will have the following API: [variable_creator_scope](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/variable_scope.py#L2395)
+*   the default implementation of the tf.Variable interface will be ResourceVariable
+    *   RefVariable will be kept in tf.compat.v1 and will be the default implementation for tf.compat.v1.Variable
+    *   tf.compat.v1.Variable will have a use_resource argument to control whether a resource variable or a ref variable will be created
+*   symbols like tf.assign* will be removed in favor of methods in tf.Variable
+    *   in tf.compat.v1 these symbols will be marked as deprecated and will call the corresponding methods in the Variable object instead
+
+
+## Detailed changes
+
+
+### tf.Variable class
+
+The tf.Variable class will be an abstract base class which defines a tf.Variable interface. Initially this interface will have enough abstract methods such that the user-visible API of tf.Variable does not change.
+
+There will be two main implementations of this interface: RefVariable, with the legacy ref edges, available only in tf.compat.v1, and ResourceVariable, which is the default for the v2 API. PartitionedVariable, MirroredVariable, _UnreadVariable, CastVariable, etc, are other implementations which are part of the core library. None of these implementations will be publicly visible, only tf.Variable will be.
+
+Constructing variables is done by calling tf.Variable(*args, **kwargs). Under the hood this will call a hierarchy of scoped constructor functions, similar to what is now done in variable_scope.variable. Each such constructor function can do some combination of:
+
+
+
+*   calling a base constructor to actually create a variable
+*   returning preexisting variables
+*   changing some arguments to the base constructor, and maybe calling it multiple times
+
+This is implemented by having a custom metaclass for tf.Variable which, when asked to construct a tf.Variable directly will call the factory functions, but when asked to construct subclasses of tf.Variable will do nothing and construct the child class.
```

Comment:
It would be good to include a justification for why the client API should be calling the constructor to an abstract base class instead of having users explicitly call the type of variable they want. This document just says "it will do this complicated thing" without saying what the rationale is.

Reply:
The goal is that the user should not have to know what type they want. For example, code called under distribution strategies might create MirroredVariables when the user calls tf.Variable. Think of tf.Variable as a factory function for which isinstance also works.